feat(transformer/react-refresh): support checking a variable whether is used in JSX elements or JSX-like calls

Dunqing opened the pull request:
Fixes: https://github.com/oxc-project/oxc/issues/11824

We have to introduce a visitor to pre-collect all bindings that are used in JSX elements or JSX-like calls, because a variable can be referenced before or after the declaration.

Boshen commented:
### Merge activity

* **Jun 25, 7:26 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Jun 25, 7:26 AM UTC**: `Boshen` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

Dunqing commented:
> @Dunqing I'm afraid I've run out of time to review this before I go away. Obviously the perf impact is unfortunate, but if it's important to get the bug fixed, please feel free to merge. We can catch up next week and see if any way to win back some of the perf.

I think of it a little bit, we can optimize this to introduce a special visitor like `expression`-only visitor or `js`-only visitor. Now I have handwritten to skip some unrelated ast.

Dunqing committed:
feat(transformer/react-refresh): support checking a variable whether is used in JSX elements or JSX-like calls (#11859)

Fixes: https://github.com/oxc-project/oxc/issues/11824

We have to introduce a visitor to pre-collect all bindings that are used in JSX elements or JSX-like calls, because a variable can be referenced before or after the declaration.
